### PR TITLE
🌱 Optmize the `isNodeDrainAllowed ` judgment logic.

### PR DESCRIPTION
**Pull request by qmloong**
**What this PR does / why we need it**:

Optmize the `isNodeDrainAllowed ` judgment logic.
This may reduce stress for isNodeDrainAllowed before list and filter machine collections.

**Which issue(s) this PR fixes** *
N/A

### CODE REVIEW

File: controllers/machine_controller.go [hunk 1]
```diff
@@ -469,22 +469,23 @@ func (r *MachineReconciler) isDeleteNodeAllowed(ctx context.Context, cluster *cl
...
-	// number of remaining control plane members and whether or not this
+	// Whether it is okay to delete the NodeRef depends on the
+	// number of remaining control plane members and whether this
 	// machine is one of them.
-	numControlPlaneMachines := len(machines.Filter(collections.ControlPlaneMachines(cluster.Name)))
-	if numControlPlaneMachines == 0 {
-		// Do not delete the NodeRef if there are no remaining members of
-		// the control plane.
-		return errNoControlPlaneNodes
+	if collections.ControlPlaneMachines(cluster.Name)(machine) {
```
Comment (fabriziopandini) [inline controllers/machine_controller.go:475]: Please correct me if I'm wrong, but this change greenlights deleting a worker node even if there are no more control plane machines in the Cluster, which is something that will lead to an error preventing a machine to be deleted. 

The current implmementation instead always proceed to delete the machine in the use case above (just skipping node deletion when not possible or not required)

Comment (vincepri) [reply controllers/machine_controller.go:475]: +1 the above logic was fixed in a few versions ago. When control plane machines aren't available, there is no way to delete Kubernetes objects in the target cluster, and we should make sure that the Machine controller doesn't get stuck

Comment (enxebre) [reply controllers/machine_controller.go:475]: I agree, this should indeed break the `no control plane members` unit test.

Comment (qmloong) [reply controllers/machine_controller.go:475]: Oh, sorry. I misunderstood this piece of code logic.

Comment (qmloong) [reply controllers/machine_controller.go:475]: So this code comment is outdated, right?

```
// isDeleteNodeAllowed returns nil only if the Machine's NodeRef is not nil
// and if the Machine is not the last control plane node in the cluster.
```

**Pull request closed without merging**